File: snake-scala-2d/src/main/scala/metacake/scala/snake/state/SetupState.scala
package metacake.scala.snake.state

import io.metacake.core.process.state.GameState
import io.metacake.core.process.state.GameState.Kind
import io.metacake.core.process.{ActionRecognizerPipe, Transition}
import io.metacake.s2d.input.keyboard.KeyboardDevice
import metacake.scala.snake.entity.{Direction, Food, Segment, Snake}
import metacake.scala.snake.input.{KeyConfiguration => conf}

class SetupState(segments: List[Segment]) extends GameState {

  override def kind(): Kind = Kind.NORMAL

  override def tick(delta: Long, pipe: ActionRecognizerPipe): Transition = {
    Transition.to(new SnakeState(new Snake(Direction.RIGHT, segments), new Food(4, 4)))
                    .withTriggers(conf.wTrigger, conf.sTrigger, conf.aTrigger, conf.dTrigger, conf.escTrigger)
                    .withBucket(KeyboardDevice.KEY, conf.wRecognizer, conf.sRecognizer, conf.aRecognizer, conf.dRecognizer, conf.escRecognizer)
  }
}

File: java-snake-2d/src/main/java/metacake/snake/state/SetupState.java
package metacake.snake.state;

import io.metacake.core.process.ActionRecognizerPipe;
import io.metacake.core.process.Transition;
import io.metacake.core.process.state.GameState;
import io.metacake.s2d.input.keyboard.KeyboardDevice;
import metacake.snake.SnakeApp;
import metacake.snake.entity.Direction;
import metacake.snake.entity.Food;
import metacake.snake.entity.Snake;

import java.util.Random;

import static metacake.snake.input.KeyConfiguration.*;

public class SetupState implements GameState {
    @Override
    public Transition tick(long delta, ActionRecognizerPipe pipe) {
        Random rand = new Random();
        Food food =  new Food(rand.nextInt(SnakeApp.WIDTH / Food.SIZE), rand.nextInt(SnakeApp.HEIGHT / Food.SIZE));
        return Transition.to(new SnakeState(Snake.initialSnake(Direction.RIGHT), food))
                .withTriggers(wTrigger, aTrigger, sTrigger, dTrigger)
                .withBucket(KeyboardDevice.KEY(), wRecognizer, aRecognizer, sRecognizer, dRecognizer);
    }
}